hmiddleware/connection_closer.go

```diff
@@ -0,0 +1,75 @@
+package hmiddleware
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/google/uuid"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/heroku/x/go-kit/metrics"
+)
+
+type connCloseKey int
+
+var connCloseID connCloseKey
+
+// ConnectionClosingContext adds a unique identifier to the context for a newly established connection. This function
+// is meant to be used as an http.Server's ConnContext field, and is required for the ConnectionClosingMiddleware.
+func ConnectionClosingContext(ctx context.Context, c net.Conn) context.Context {
+	return context.WithValue(ctx, connCloseID, uuid.New().String())
+}
+
+// ConnectionClosingMiddleware returns middleware that will add a "Connection: close" responses header after a single
+// connection has produced maxRequests http requests. The cacheSize determines the size of the LRU cache which tracks
+// connection request counts. This middleware requires the http.Server sets ConnectionClosingContext for it's
+// ConnContext field.
+func ConnectionClosingMiddleware(mp metrics.Provider, maxRequests, cacheSize int) func(http.Handler) http.Handler {
+	var (
+		closed = mp.NewCounter("server.connection.closes.total")
+		mtx    sync.Mutex
+	)
+
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
Should this use a passed-in logger or return an error if there's nothing else to log?

Reply:
good call, i was trying to avoid requiring a logger here, so i can just add the `error` to the return values.